Add support for Multipart requests in RESTEasy Reactive

geoand opened the pull request:
Fixes: #14601

As mentioned in the title, this only adds support for reading multipart form data - which seems to be the most common case.

The idea here is to use code similar to what Vert.x uses with its `BodyHandlerImpl` to populate the `RoutingContext` and `HttpServerRequest` with the appropriate data from the multipart body, and then use the `__quarkus_rest_inject` mechanism available to parameter handling, in order to call a generated piece of code that populates the POJO by pulling data from the context.
The user facing API (`@Multipart` and `@PartType`) is pretty much the same as what is available in RESTEasy Classic.

One thing this PR does not do is provide an abstraction for the file upload handling that will work for non Vert.x enviroments

geoand committed:
Add support for Multipart requests in RESTEasy Reactive

Fixes: #14601

geoand commented:
PR updated to use the commit from https://github.com/quarkusio/quarkus/pull/14880 as the first commit.

geoand commented:
Something seems to have gone wrong when I updated the PR.
I'll check it later

geoand commented:
> Something seems to have gone wrong when I updated the PR.
> I'll check it later

Should be fixed now

stuartwdouglas commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/multipart/FileUpload.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * @return the content transfer encoding of the upload - this describes how the upload was encoded in the form submission.
+     */
+    String contentTransferEncoding();
```
We should not be exposing this, its a protocol level detail that is not really relevant.

stuartwdouglas commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/multipart/FileUpload.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * @return the actual temporary file name on the server where the file was uploaded to.
+     */
+    Path uploadedFile();
```
This assumes everything is a file, which is often not the case. If there is no associated file name then this is just a normal file upload (e.g. a text box). IMHO we should only be creating files for the actual file parts of the request, and provide a convenient way to get the non-file parts.

geoand replied:
In light of my comment below, do you think this still needs to be addressed?

stuartwdouglas replied:
Ah, in that case I think it is fine.

stuartwdouglas commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/MultipartFormHandler.java:
```diff
@@ -0,0 +1,221 @@
+                    if (handleFileUploads) {
+                        // we actually upload to a file with a generated filename
+                        uploadCount.incrementAndGet();
+                        String uploadedFileName = new File(uploadsDirectory, UUID.randomUUID().toString()).getPath();
```
This will create a file for every form field, so e.g. if you have a form with a checkbox and a file upload the results of the checkbox will be written out to disk as a file. I think we need to be a bit more selective here, and only write out parts that have a filename set.

geoand replied:
Actually that's not the case, as Vert.x already handles this under the hood.
The `uploadHandler` is only invoked for parts of the HTTP body that contain `filename`.
The rest are simply added to `request.formAttributes()`.

I'll update the test case to make that clear

stuartwdouglas commented:
Even though it is not critical for this PR I think we really need our own mutipart/form parsing support to support multiple transports. I have one in Undertow that we can probably use as a base.

https://github.com/quarkusio/quarkus-http/blob/master/core/src/main/java/io/undertow/server/handlers/form/MultipartParser.java

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/MultipartFormHandler.java:
```diff
@@ -0,0 +1,221 @@
+
+            context.request().setExpectMultipart(true);
+            if (handleFileUploads) {
+                makeUploadDir(context.vertx().fileSystem());
```
Good point, fixed

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/MultipartFormHandler.java:
```diff
@@ -0,0 +1,221 @@
+    public void handle(ResteasyReactiveRequestContext context) throws Exception {
+        // in some cases, with sub-resource locators or via request filters,
+        // it's possible we've already read the entity
+        if (context.hasInputStream()) {
```
I don't think we are doing that for form-encoded either ATM.

Perhaps we kick it down the road? It doesn't sound like a very common use case :)

stuartwdouglas replied:
Yea, eventually I want to add our own Multipart support so it is not so backend dependent, we can handle it then.

geoand replied:
👍🏼

geoand commented:
> Even though it is not critical for this PR I think we really need our own mutipart/form parsing support to support multiple transports. I have one in Undertow that we can probably use as a base.
> 
> https://github.com/quarkusio/quarkus-http/blob/master/core/src/main/java/io/undertow/server/handlers/form/MultipartParser.java

I agree that it would be a good follow up.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:
```diff
@@ -225,4 +228,42 @@ public ServerRestHandler get() {
+                                .orElse(true);
+                        deleteUploadedFilesOnEnd = ConfigProvider.getConfig()
+                                .getOptionalValue("quarkus.http.body.delete-uploaded-files-on-end", Boolean.class)
+                                .orElse(false);
```
I'm not sure this should be an option, but at the very least it should default to `true` and force people to decide what to do with the uploads: move them someplace they take care of them, like dedicated FS, or cloud storage or DB, whatever, but away from our responsibility. Otherwise these will persist on application restart and grow forever without the user noticing.
IMO it's better for users to notice that we removed their files right away during dev (and document it) and decide what they need.

geoand replied:
Well, we already have this option in Quarkus so I just reused it.

I'm all for making this option `true` be default in Quarkus

FroMage replied:
Oh, I didn't realise that. I would say yes, but it's possible that existing users depend on this behaviour then. I don't think we should follow that behaviour in RR.

geoand replied:
👍🏼

FroMage reviewed (changes requested):
Sorry, I still don't understand why this is not similar to how we handle `@BeanParam`:

- injector method inside the class itself (not a static method in a generated class)
- one injector method per hierarchy level, allows access to private fields
- no need to care about setters

I know you mentioned it was different to `@BeanParam` but IMO the two annotations could actually be merged as one meaning "stuff whatever you can in there" and then the presence of form or other params in there would tell us if we need to register additional handlers (only for multipart form, I think).

Even if we don't support mixing multipart and bean params in the same class, I don't see how their injector implementations need to be different?

FroMage commented:
Also, this will need documentation :)

geoand commented:
The idea of how you handle `@BeanParam` is similar (in the sense that you generate a `_quarkus_rest_inject` method), but if you look at what can be part of `@MultipartForm` and how it's obtained, it's totally different from what can be part of `@BeanParam`.
Apart from reusing the mechanism you put in place that I mentioned above, I see absolutely nothing else that can be shared.

geoand commented:
> Even if we don't support mixing multipart and bean params in the same class, I don't see how their injector implementations need to be different?

Apart from the fact they inject different things, the reason I went for Gizmo instead of ASM is that it is SOOOOOO much easier to write and maintain.
If anything, I would propose moving the `@BeanParam` injector to Gizmo instead of moving what I did here to ASM.

FroMage commented:
> If anything, I would propose moving the @BeanParam injector to Gizmo instead of moving what I did here to ASM.

I would have said sure, but the reason it's in ASM is because I generate one injector method per class hierarchy level, which allows us to access private fields. Otherwise you can't.

geoand commented:
> > If anything, I would propose moving the @BeanParam injector to Gizmo instead of moving what I did here to ASM.
> 
> I would have said sure, but the reason it's in ASM is because I generate one injector method per class hierarchy level, which allows us to access private fields. Otherwise you can't.

Why not? If you look at the tests, my Gizmo code can use either `public` fields or `public` setters. This is a typical pattern we have in a lot of places.
I admit thought that I might have misunderstood your point :)

FroMage commented:
Yeah, I'm saying it doesn't support `private` or `package-protected` fields, which is the pattern we're telling people to use, for ArC, for example.

I'm all for supporting and using `public` fields, don't get me wrong. But supporting all the fields is even better ;)

geoand commented:
In that case my preference would be to just strip the `private` modifier with ASM and continue doing what I did.
IMHO, Gizmo makes code generation code manageable, whereas ASM is just too much of a pain and I would like to avoid it as much as possible.

FroMage commented:
If you strip the modifier, you can use package-protect to get to the fields but only if you split the injectors into same-package one-injector-class-per-hierarchy-class to allow them package access, like ArC does.

Don't get me wrong, I'm all for reducing ASM usage. But if you do that, also do it for bean params, because aside from the particulars about specific field types, they do the same thing :)

FroMage commented:
Well, I guess one reason I wanted to avoid this was to also avoid creating extra classes for no good reason…

geoand commented:
> If you strip the modifier, you can use package-protect to get to the fields but only if you split the injectors into same-package one-injector-class-per-hierarchy-class to allow them package access, like ArC does.
> 

Yeah, the generated class is in the same package, so indeed this will work.

> Don't get me wrong, I'm all for reducing ASM usage. But if you do that, also do it for bean params, because aside from the particulars about specific field types, they do the same thing :)

I do plan to replace the bean param ASM stuff with gizmo, just not now :)

geoand commented:
> wanted to avoid thi

Yeah, I completely understand that. But from a maintenance perspective, ASM is nightmare... I would rather have something a little more maintaneable at the cost of a tiny amount of extra classes.

FroMage commented:
> Yeah, the generated class is in the same package, so indeed this will work.

Only if you generate one class per hierarchy level, each in the same package as the class you're injecting to. I had to deal with all that to pass the TCK, that's why I said this was exactly the same problem and could have the same solution.

FroMage commented:
Consider the case of two classes:

```java
package foo;

public class Foo {
 String something;
}

package bar;

public class Bar extends Foo {
 String other;
}
```

FroMage commented:
> cost of a tiny amount of extra classes

I'm not sure there's ever "a tiny amount of extra classes" there's always "oh well, lots of tiny amounts end up with a huge number of classes and now the startup time is affected", no?

geoand commented:
> > cost of a tiny amount of extra classes
> 
> I'm not sure there's ever "a tiny amount of extra classes" there's always "oh well, lots of tiny amounts end up with a huge number of classes and now the startup time is affected", no?

Given that this feature isn't too prevalent, I would say there is no danger. I mean, I don't see applications having 100 Multipart POJOs.
And if they do and it becomes a problem, I would rather deal with it then instead of having to maintain complex ASM from the start.

FroMage commented:
I get your points, but I worry that we now have two different implementations of essentially the same thing, one of which is hard to maintain, the other which doesn't work in all cases and creates extra classes.

I started with the second strategy, but changed it at the time, thinking runtime perf and completeness for users is more important than maintenance for us.

geoand commented:
> I get your points, but I worry that we now have two different implementations of essentially the same thing, one of which is hard to maintain, the other which doesn't work in all cases and creates extra classes.
> 

I just pushed an update that shows that the generated code handles the cases you mention. 
So we are not in the situation where the second implentation doesn't handle some cases.

> I started with the second strategy, but changed it at the time, thinking runtime perf and completeness for users is more important than maintenance for us.

If this were a core feature, I would completely agree. But let's face it. neither `@BeanParam` nor `@MultipartForm` are core features...
From all the use cases we have seen from users, rarely (if ever) are they used multiple times in the same application. So having O(n) number of generated classes I think will never become a problem

FroMage commented:
OK fine, but we still need to figure out what to do about deleting files. And mention it in the docs, as it's pretty important IMO.

geoand commented:
> OK fine, but we still need to figure out what to do about deleting files. And mention it in the docs, as it's pretty important IMO.

Right.

I completely agree about changing the default about deleting files.
I'll add a second commit to the PR doing that and ping the folks that should know about it.

geoand committed:
Change default file upload settings

Change quarkus.http.body.delete-uploaded-files-on-end to true
and quarkus.http.body.uploads-directory to use tmp dir

This is done because for one it doesn't make sense to accumulate uploaded files by default
and also it the tmp dir should be used instead of the working directory as some applications
might not be able to write to the current working directory

geoand commented:
@cescoffier @gsmet the second commit of this PR changes the default value of `quarkus.http.body.delete-uploaded-files-on-end` to `true` as it doesn't make sense to keep uploaded files around as the default setting.

Do you agree?

FroMage commented:
Note that this comes from my own experience running out of directory entry buckets for long-running applications. Doing a `listdir` on the damn `file-uploads` folder took minutes because it had thousands of unneeded old files.

geoand commented:
> Note that this comes from my own experience running out of directory entry buckets for long-running applications. Doing a `listdir` on the damn `file-uploads` folder took minutes because it had thousands of unneeded old files.

You are absolutely right. We should never default to keeping files around....

geoand commented:
I fixed an issue with the 2nd commit

geoand commented:
@stuartwdouglas @FroMage can I get re-review of this please?

Thanks

FroMage reviewed (approved):
LGTM

geoand commented:
CI is green 🥳

gsmet commented:
Agreed on `delete-uploaded-files-on-end`.

@stuartwdouglas I let you have another look? It would be great to have it in CR1, even if small adjustments are required.

stuartwdouglas commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:
```diff
@@ -225,4 +228,42 @@ public ServerRestHandler get() {
+                        // TODO: reevaluate this
+                        handleFileUploads = ConfigProvider.getConfig().getOptionalValue("quarkus.http.body.handle-file-uploads",
+                                Boolean.class).orElse(true);
+                        uploadsDirectory = ConfigProvider.getConfig().getOptionalValue("quarkus.http.body.uploads-directory",
```
I think there are a few problems here:
- This is runtime config being read in static init, so it will appear to work in JVM mode but won't work as expected in native mode.
- Currently this config only applies to the Vert.x body handler, which RESTEasy Reactive does not use. I don't know if we really want to overload this configuration.
- handle-file-uploads is not a good config option imho, the fact that the user has defined a multipart resource that is expecting a file is an indication that they want to handle file uploads. We should accept files on multipart endpoints, and for other endpoints leave it up to the MessageBodyReader how to handle it. 
- The default upload directory should be System.getProperty("java.io.tmpdir"), not 'file-uploads', as some apps won't have permission to write to the current working directory.
- merge-form-attributes makes no sense in this context, it just adds the form params to the query params, so will cause problems for us if you have a form param and a query param of the same name (basically you will be able to use @QueryParam to inject what should be an @FormParam). 
- The max request size is available under io.quarkus.vertx.http.runtime.VertxHttpRecorder#MAX_REQUEST_SIZE_KEY, and for fixed length requests is already enforced. I think our impl should just use this to avoid the static init config issue.

geoand replied:
The problem with `quarkus.http.body.uploads-directory` I think is that because of its name, users will expect it to work properly for all HTTP stacks. I could certainly remove it from here, but wouldn't it be confusing for users that such a "global" HTTP property doesn't work?
About `quarkus.http.body.handle-file-uploads`, I certainly agree but again we have the problem that "global" HTTP configuration will appear to not work in RESTEasy Reactive. Are we OK with that?

geoand replied:
I overhauled how this configuration is passed and removed the fileds you mentioned that are non necessary.

However, I haven't changed the default values and still rely on Quarkus defaults.
Let me know if you still think we should not abide by the Quarkus "global" http settings

FroMage replied:
FTR, I do agree that an app should have a single "tmp" folder where it puts stuff. If that's configured with `java.io.tmpdir` then fair enough, but perhaps `quarkus.http.body.uploads-directory` should default to `{java.io.tmpdir}/uploads` then?

geoand replied:
That makes a lot of sense. Let me see if Quarkus config allows that

geoand replied:
OK, I changed the default to use the tmp dir approach.

cc @gsmet @cescoffier

geoand replied:
@stuartwdouglas I think I addressed all of your concerns

stuartwdouglas commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartSupport.java:
```diff
@@ -0,0 +1,116 @@
+
+        Class<?> type = loadClass(typeClassStr);
+        Type genericType = type;
+        if (signature != null) {
```
Not a blocker, but shouldn't this be parsed in the static init block of the generated class and stored in a field, same with the media type?

geoand replied:
Good idea, I'll do that

geoand replied:
Done
